### DISCUSSION TITLE
TUI Key Binding/Keymap Crate

### DISCUSSION THREAD

**Discussion by thebest132** [ratatui/ratatui#627] [Discussion]
Hi,

I'm currently working on a keymap crate to make it easier for people to handle key presses in their TUI apps.

Here is what it will do:
- Key sequences (e.g. press "a" then "b" then "c" in succession)
- Profiles
    - Think how different Vim modes have different keys binded
    - Or think how maybe in the main menu of an app you would have certain key bindings and different keys binded in other parts of the app
- (Possible) support for external .toml configuration, however this might be too specific to the app
- Easily bind key actions/sequences to functions

First, I would like to know if this is a good idea, and if any of you would find this useful. Second, I would like to know how you guys would prefer to set the keymap. [Here is a file](https://github.com/taahz/tui-key-bindings/blob/master/src/need_feedback.rs) that lists all the syntax and options.

Thanks!

**Comment by joshka** [comment]
It's definitely a good idea and one I definitely want to happen. For some more inspiration, take a look at https://crates.io/crates/crokey (which you can see in action configuring bacon in bacon.toml) at https://github.com/ratatui-org/ratatui/blob/edacaf7ff4e4b14702f6361af5a6da713b7dc564/bacon.toml#L94-L100

Also https://crates.io/crates/modalkit and https://crates.io/crates/tui-textarea (see https://github.com/rhysd/tui-textarea/issues/51) have some ideas in this and it's worth searching the discord channels for mentions of binding too.
- https://discord.com/channels/1070692720437383208/1080107705672413264/1154207800294113331
- https://discord.com/channels/1070692720437383208/1080107705672413264/1156938009762086952
- and others - search "binding" / "keybinding"

Profile might be also called Mode / Context / Layer in some other apps. VScode does [conditional contexts ](https://code.visualstudio.com/docs/getstarted/keybindings#_when-clause-contexts) (which is probably unnecessary for small TUI apps, but might be worth considering as something to help constrain your existing design and have as a stretch goal for later).

For a Ratatui app that has something already, see yazi (starting point is https://github.com/sxyazi/yazi/blob/main/yazi-config/src/lib.rs#L39).

You'll note that a lot of these have different ways to get to the same thing - so finding the "right" way might take some exploration.

The use cases I'd love to see for something like this would be:
1. There is some way to specify a default configuration
2. The user can easily add configuration in TOML that allows for the defaults to be overriden
3. Model easily lends itself to displaying the bindings to the user in a help window (which likely entails both individual bindings and groups of bindings)
4. Supports all the backends we support - Crossterm/Termion/Termwiz
5. Does not make assumptions about specific application approaches
6. Might be worth considering making it non-Ratatui specific (i.e. more generally useful for tuis, regardless of whether they use Ratatui or not)

Pinging @kdheepak @pm100 @rhysd who have variously thought about some of these issues in the past / present.

**Reply by thebest132** [reply]
Thanks for the response and those crates! Do you think this is worth making? Is there anything else that already does this? I'm not really sure if the crokey crate is meant for this purpose or just for easy KeyEvents.

1. By this you mean in Rust in the TUI app right, like a macro_rules! macro?
2. Will do that
3. So easily convert the bindings into strings to be used in the app?
4. Will do that
5. I'll look at multiple sources to make sure of this
6. Yes, the reason I called it tui-key-bindings like the others is because when I looked at the tui-input crate, it didn't seem ratatui specific. If you think this crate should not be listed in the README.md of ratatui, I'd love to hear some alternative name ideas (it seems tui-* is used only in tui/ratatui specific crates?). If it is, let me know if tui-key-bindings, tui-keymap or something else is better.

**Reply by joshka** [reply]
>Do you think this is worth making?

Yes. Absolutely - Rust needs a generic crate that works for this so we can make backend agnostic apps that handle keybindings well.

>I'm not really sure if the crokey crate is meant for this purpose or just for easy KeyEvents.

Crokey is just the serialization part of this and is crossterm specific. This idea will be the part that goes past that, but probably should consider serialization as a goal.

> 1. By [setting defaults] you mean in Rust in the TUI app right, like a macro_rules! macro?

I do like how yazi does this with having files that are included for this, but either way.

> 3. So easily convert the bindings into strings to be used in the app?

Yeah - this requirement mainly dictates that there's some sort of string representation of the keys and the actions that are bound.

> 6. Yes, the reason I called it tui-key-bindings like the others is because when I looked at the tui-input crate, it didn't seem ratatui specific. If you think this crate should not be listed in the README.md of ratatui, I'd love to hear some alternative name ideas (it seems tui-* is used only in tui/ratatui specific crates?). If it is, let me know if tui-key-bindings, tui-keymap or something else is better.

I kinda like the idea of crates just using `tui-` for their name - makes them easy to find / group. There's not really a consistency on namespacing across all crates, so the answer is do what works.

**Comment by joshka** [comment]
Just now, I slapped a commit on some code I had lying around on my local dev folder, and pushed it to https://github.com/joshka/keybinding and published it to https://crates.io/crates/keybinding to reserve the crate name.

**Reply by thebest132** [reply]
Oh, I missed this when writing my other response. I had no idea that name was available, too. Should I still work on this in my own repository?

**Reply by joshka** [reply]
Yeah keep going in your repo - don't worry about the crate I made here - I mainly grabbed it in case we do get to the point where there is a generic lib that solves this, but It's not something I am focusing deeply on right now.

**Comment by kdheepak** [comment]
In `ratatui-async-template` I implemented a way to read a `Vec<KeyEvent>` from disk and mapped that to an action.

https://github.com/ratatui-org/ratatui-async-template/blob/8bebe2129f4e97dc50340bd7fa77822250a17c9a/template/src/config.rs#L83

Just posting as a reference. I absolutely agree that a generic crate to handle this well will go a long way to making more configurable TUI apps with ratatui.

**Reply by thebest132** [reply]
Looking at the `Config::new` function, it seems config is really specific to the app and I shouldn't put anything like that in the keybinding crate. This is one of the reasons I don't necessarily want to have a specific file just for users of this crate. I think it could be easier to set the bindings in Rust itself, since those probably won't be changed often, as opposed to user-defined configs. How do you think keybindings should be set in Rust (e.g. custom `macro_rules!` macro like [this](https://github.com/helix-editor/helix/blob/40959bb449ec9dfbff9249332853a02b50f483fe/helix-term/src/keymap/default.rs#L8C1-L8C1)). I need input on how this crate should be designed to be ideal for the user.

**Reply by joshka** [reply]
A step back question might be why define the defaults in rust? (more generally if something is **data**, should it be configured or programmed?)

**Reply by thebest132** [reply]
I would answer configured to this. But I'll have to figure out how to nicely have both the app's default config and a user config. The thing is, having this kind of thing outside of the code sort of says to the user it's meant to be changed, which you don't want for the defaults. To stop this, you would do something like yazi which is have a separate folder for the presets. For certain apps this feels unnecessary though, and you'll have a config folder with one toml and a presets folder with one toml. Putting the config in a separate file feels more like forcing the user to structure their app a certain way that might not work for all. Obviously it's not actually too bad, I'm just making a point. How would you do this? Preset and config folder? A `keybindings.toml` file in the root and a config folder? Both the presets and default in one config folder? Something else? I don't want to make any bad design decisions.

**Reply by joshka** [reply]
One possible trick is to use [include_str!](https://doc.rust-lang.org/std/macro.include_str.html) to compile the keybindings.toml file into the app, so the user doesn't ever see that file, and then use [toml!](https://docs.rs/toml/latest/toml/macro.toml.html) to convert that string into an typed value. You could then use that as your defaults passed into a config from https://docs.rs/config/0.13.3/config/.

The reason for asking this question like this is that it gets you out of the mindset of how to represent the API in a way that works nicely for a dev trying to call into the  lib, because TOML is probably good enough for that part instead of writing your own DSL with macros. Then you just instead need to worry about the types and the relationships between them. You've probably got a `KeyCode` type (or a `KeyChord` / ?) and some Action type (perhaps it's generic?), and a way of specifying some sort of modality or context - together those things are a `KeyBinding`...

Something like the following would be the app's perspective on using this:

```toml
bindings = {
  "ctrl+a" = "action::beginning_of_line",
}
```

```rust
let default_bindings = toml!(include_str!("default_keybindings.toml")).bindings;
let config = Config::builder()
  .set_default("bindings", default_bindings)
  .add_source(File::with_name("myapp-keybindings")
  .build();
let bindings = config.get::<Vec<KeyBinding>>();
dbg!(bindings);
```

```rust
vec![
  KeyBinding {
    key_code: KeyCode {
      ...
    },
    action: Action::Foo,
    context: ...
  },
  KeyBinding { ... },
  KeyBinding { ... };
];
```

There's probably a missing middle in there somewhat (something that's incredibly difficult to actually achieve with the bits available. Maybe not though, and if so that would be great.

**Comment by thebest132** [comment]
It's been a while, wow. I haven't programmed for a long time. But I just remembered about this. I thought of something else. What if we had something like this:

```rust
#[context]
struct MyApp {
    #[var("programInfoVariable")]
    program_info_variable: u32
}

// ...

#[bind(key = "a", when = "programInfoVariable == 5")]
fn my_function(context: MyApp) {
    context.program_info_variable += 1;
}
```

Thought this up randomly. I kind of liked it. It could also be in a JSON file where the proc macro shows what the function name is in the JSON file:

```json
{ "key": "a", "command": "myFunction", "when": "programInfoVariable == 5" }
```

```rust
// (struct stays the same)

#[name("myFunction")]
// (function here)
```

Let me know what you think. Thought of this after looking into Helix. There could be problems with this in large scale apps.

**Reply by joshka** [reply]
That's a neat idea. I like the bind macro there, but the context feels like it's probably unnecessary.
What would the macro generated code look like there?

**Reply by thebest132** [reply]
I'm not sure about the macro generated code right now, I haven't tried it. But I'm not sure what you mean by the context being unnecessary. Every function needs to have some data. I can see how it might not be completely central though. I took a look at Lapce, and it doesn't seem to have a central context. Instead, the functions are across many different structs. However, the code is pretty convoluted, and after looking for a while I came to a dead end with the GitHub search not showing anything and I have no idea how they actually call those functions. That doesn't seem viable for a generic crate.

I did see, though, that Lapce does some things similarly. They use the strum crate to create the "when" context variables which is basically the same as what we're doing here. They have enums as the commands, with the strum crate used to set their function names. What confused me is how they actually call those functions. It looked like it had something to do with a trait for all the structs where functions are called but, again, I reached a dead end. I'm not the smartest so maybe you know what that's about.

**Reply by thebest132** [reply]
I looked more into Lapce and saw it's too concrete to take inspiration from. I tried looking into Yazi too and it's too complex of a codebase for me to understand. There was an event system and router and stuff. Strangely, Helix seemed to be the simplest even though it's the largest project. They are the ones I took inspiration from with this "context" stuff. It's how they do it. Each function is not in a struct and has one parameter which is the context.

**Reply by joshka** [reply]
> I'm not sure about the macro generated code right now, I haven't tried it.

Something to remember is that a macro is just another name for code that generates code. The macro `#[bind(key = "a", when = "programInfoVariable == 5")]` compiles down into some concrete code. You can use cargo expand to see what that looks like (or the equivalent rust-analyzer command in your IDE of choice)

> But I'm not sure what you mean by the context being unnecessary. Every function needs to have some data. I can see how it might not be completely central though.

I meant that I'm not sure that I fully understand the need to tag the MyApp struct with a macro - i.e. what that does / helps
